webpack.config.js

```diff
@@ -0,0 +1,110 @@
+
+function getPlugins(env, argv) {
+  const plugins = [
+    new webpack.IgnorePlugin(/^esprima$/, /js-yaml/), // Ignore Esprima import for js-yaml
+    new webpack.IgnorePlugin(/^\.\/locale$/, /moment$/), // Ignore all optional deps of moment.js
+    new webpack.DefinePlugin({
+      NETLIFY_CMS_VERSION: JSON.stringify(`${ pkg.version }${ env.production ? '' : '-dev' }`),
+    }),
+    new webpack.NoEmitOnErrorsPlugin(), // Default for production mode, but adding to development mode
+  ];
+  if (env.production) {
+    // Extract CSS
+    plugins.push(new MiniCssExtractPlugin({
+      filename: "[name].css",
+    }));
+    // During beta phase, generate source maps for better errors
+    plugins.push(new webpack.SourceMapDevToolPlugin({
+      // asset matching
+      test: /\.js?$/,
+
+      // file and reference
+      filename: '[file].map',
+
+      // don't include source file content, since we link to the actual file
+      noSources: true,
+
+      // sourcemap is in 'dist', webpack context is in 'src'
+      moduleFilenameTemplate: info => path.posix.normalize(`../src/${ info.resourcePath }`),
+    }));
+  }
+  if (env.development && env.write) {
+    plugins.push(new WriteFilePlugin());
+  }
+  return plugins;
+}
+
+module.exports = function(env, argv) {
+  const plugins = getPlugins(env, argv);
+
+  return {
+    mode: env.production ? 'production' : 'development',
+    entry: {
+      cms: './index',
+    },
+    output: {
+      path: path.join(__dirname, 'dist'),
+      filename: '[name].js',
+      library: 'netlify-cms',
+      libraryTarget: 'umd',
+      umdNamedDefine: true,
+    },
+    context: path.join(__dirname, 'src'),
+    module: {
+      noParse: /localforage\.js/,
+      rules: [
+        {
+          test: /\.js?$/,
+          use: 'babel-loader',
+          exclude: /node_modules/,
+        },
+        {
+          /**
+           * CSS loader for npm modules that are shipped with CSS that should be loaded without processing.
+           * List all of theme in the array
+           */
+          test: /\.css$/,
+          include: [/redux-notifications/],
+          use: [
+            (env.production ? MiniCssExtractPlugin.loader : 'style-loader'),
+            'css-loader',
+          ],
+        },
+        {
+          /* We use PostCSS for CMS styles */
+          test: /\.css$/,
+          exclude: [/node_modules/],
+          use: [
+            (env.production ? MiniCssExtractPlugin.loader : 'style-loader'),
+            {
+              loader: 'css-loader',
+              options: {
+                importLoaders: 1,
+              },
+            },
+            { loader: 'postcss-loader' },
+          ],
+        },
+        {
+          test: /\.svg(\?v=\d+\.\d+\.\d+)?$/,
+          exclude: [/node_modules/],
+          loader: 'svg-inline-loader',
+        },
+      ],
+    },
+    plugins,
+    target: 'web', // Make web variables accessible to webpack, e.g. window
+    devServer: {
```

Comment:
Sorry, I guess I don't understand why the server change was necessary? I'm definitely not opposed to switching it back now that the WriteFile plugin has been upgraded, I just want to understand why.

Reply:
I was getting a schema error on the settings using `server`:
```
× ｢wds｣: Invalid configuration object. Webpack has been initialised using a configuration object that does not match the API schema.
 - configuration has an unknown property 'server'. These properties are valid:
   object { mode?, amd?, bail?, cache?, context?, dependencies?, devServer?, devtool?, entry?, externals?, loader?, module?, name?, node?, output?, optimization?, parallelism?, performance?, plugins?, profile?, recordsInputPath?, recordsOutputPath?, recordsPath?, resolve?, resolveLoader?, stats?, target?, watch?, watchOptions?arallelism?, performance?, plugins?, profile?, recordsInputPath?, recordsOutputPath?, recordsPath?, resolve?, resolveLoader?, stats?, target?, watarallelism?, performance?, plugins?, profile?, recordsInputPath?, recordsOutputPath?, recordsPath?, resolve?, r
arallelism?, performance?, plugins?, profile?, recordsInputPath?, recordsOutputPath?, recordsPath?, resolve?,
arallelism?, performance?, plugins?, profile?, recordsInputPath?, recordsOutputPath?, recordsPath?, resolve?, resolveLoader?, stats?, target?, watch?, watchOptions? }
   For typos: please correct them.
   For loader options: webpack >= v2.0.0 no longer allows custom properties in configuration.
     Loaders should be updated to allow passing options via loader options in module.rules.
     Until loaders are updated one can use the LoaderOptionsPlugin to pass these options to the loader:
     plugins: [
       new webpack.LoaderOptionsPlugin({
         // test: /\.xxx$/, // may apply this only for some modules
         options: {
           server: …
         }
       })
     ]
error An unexpected error occurred: "Command failed.
Exit code: 1
```

Reply:
It should be `serve`, not `server`. Anyway, it [doesn't support function-based configs](https://github.com/webpack-contrib/webpack-serve#webpack-function-configs) yet, so lets just go with `dev-server` for now.

Reply:
> Due to some special preprocessing that is part of webpack-cli, and not webpack proper, webpack-serve cannot fully support configs that export a function without creating a bit of a mess. We're working on a solution for this, but for the moment we don't recommend using webpack-serve with configs that export a function()

Right, this is the official quote. We can upgrade once they have it stable